Add plugin adapters.

Cybermaxke opened the pull request:
[API](https://github.com/SpongePowered/SpongeAPI/pull/1990) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/2238) | **Forge** | Vanilla | [KotlinAdapter](https://github.com/Cybermaxke/KotlinAdapter)

Cybermaxke committed:
Add plugin adapters.

Cybermaxke committed:
A few improvements.

Cybermaxke committed:
Add plugin injection modules.

Cybermaxke committed:
The adapter needs to handle plugin modules.

Singletons get initialized when the injector is constructed, at this point required plugin modules may still be missing, so that's why this should be added at the same point as the plugin instance module.

Cybermaxke committed:
Update for moved default adapter.

Cybermaxke committed:
Throw a exception if the plugin instance isn't a singleton.

Cybermaxke committed:
Load the adapter and module classes with the mod class loader.

XakepSDK commented on src/main/java/org/spongepowered/mod/plugin/SpongeModPluginContainer.java:
```diff
@@ -273,15 +280,44 @@ public void constructMod(FMLConstructionEvent event) {
+            if (injectionModuleTypes != null && injectionModuleTypes.size() > 0) {
+                for (Type injectionModuleType : injectionModuleTypes) {
+                    final Class<?> moduleClass = Class.forName(injectionModuleType.getClassName(), true, modClassLoader);
+                    modules.add((Module) moduleClass.newInstance());
```
Create it using injector, so modules can depend on sponge injectable things

XakepSDK commented on src/main/java/org/spongepowered/mod/plugin/SpongeModPluginContainer.java:
```diff
@@ -273,15 +280,44 @@ public void constructMod(FMLConstructionEvent event) {
+            }
+
+            final PluginAdapter adapter = adapterClass == PluginAdapter.Default.class ? DefaultPluginAdapter.INSTANCE :
+                    (PluginAdapter) adapterClass.newInstance();
```
Inject? But i'm not sure about here

Cybermaxke replied:
I will create the instance with the parent injector, adapters should be reusable for multiple plugins. Although it's currently not implemented this way.

Cybermaxke committed:
Create the modules and adapter using a injector.

Cybermaxke committed:
More work

Cybermaxke committed:
Plugin instance construction no longer uses child injectors.

There are problems that implicit injections (JIT) are being handled by the parent injector, which causes all child provided bindings to be ignored. This works around that limitation by copying the bindings of a previous module/injector.

Cybermaxke committed:
Move some code to common.

ImMorpheus commented:
https://github.com/SpongePowered/SpongeAPI/pull/1990#issuecomment-650779079
